Transition to west

carlescufi opened the pull request:
Required changes to transition to using `west` instead of the built-in version, which is removed by this series.

Feel free to push to my branch or comment on it.

carlescufi committed:
scripts: remove west from scripts/

west will now be installed via pip in order for the bootstrapper to be
decoupled from the west runners.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
cmake: Use west installation directly

Since we install west from pip and it becomes an executable, use it
directly assuming it's in the user's PATH.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: scripts: Retrieve west path at runtime

Since west is no longer part of the Zephyr tree, we can no longer
hardcode its path anymore. Instead, use west itself to retrieve its
path, in order to point the documentation build to the correct
active west installation.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Use west instead of Git to clone Zephyr

Since west is no longer included in the Zephyr repository, instruct
users to install and use west in order to obtain the Zephyr source code.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

nashif commented on cmake/flash/CMakeLists.txt:
```diff
@@ -89,9 +89,7 @@ foreach(target flash debug debugserver attach)
-    PYTHONPATH=${ZEPHYR_BASE}/scripts/meta
-    ${PYTHON_EXECUTABLE}
-    ${ZEPHYR_BASE}/scripts/meta/west/main.py
+    west
```
should we maybe use find_program() here?

carlescufi replied:
I thought about it, but did not want it to be mandatory. I will add a check elsewhere.

SebastianBoe replied:
find_program should be used.

Using it does not imply it is mandatory.

Using find_program would be consistent with other host tool dependencies and would give the user a consistent way to override it's location (-DWEST=/path/to/west).

SebastianBoe commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,14 +20,42 @@ Follow one of the following guides for your host operating system.
+Get the source code
+*******************
+
+In order to get the source code for the Zephyr Project you need to use
```
Using the word 'need' communicates that it is not possible to manually clone the git repo's, which I assume is not so.

dbkinder replied:
How about:

    Zephyr's multi-purpose :ref:west` tool lets you easily get the Zephyr project source code,
    instead of manually cloning the Zephyr repos.

dbkinder commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,14 +20,42 @@ Follow one of the following guides for your host operating system.
+==============
 
-To clone the Zephyr source code repository from GitHub:
+Install the bootstraper for Zephyr's command-line tool, `west <west>` in a
```
I think you're trying to link to the west documentation, so:

    `west <west>`

should be as was done earlier:

    :ref:`west`

dbkinder commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,14 +20,42 @@ Follow one of the following guides for your host operating system.
+Clone the Zephyr Repositories
+=============================
+
+To clone the Zephyr source code repositories from GitHub:
```
```suggestion
Clone the Zephyr source code repositories from GitHub using the ``west`` tool:
```

dbkinder commented on doc/getting_started/installation_win.rst:
```diff
@@ -196,13 +196,12 @@ to set it up:
-Since you're using MSYS2, when you're cloning Zephyr in the next step of the
-guide, use this command line instead (i.e. add the ``--config
-core.autocrlf=false`` option).
+Since you're using MSYS2, before you clone Zephyr set the ``autocrlf`` option
```
Another PR (#11300) is removing MSYS2 support (and documentation), so we shouldn't mention MSYS2 here.

dbkinder commented on doc/release-notes.rst:
```diff
@@ -12,12 +12,12 @@ versions of the Zephyr project are released approximately every three-months.
 `GitHub tagged releases`_ page corresponding to each release), or use
-Git clone and checkout commands, such as:
+Git checkout commands, on an existing Zephyr installation obtained using the
+`west tool <_get_the_code>`.
```
That's not going to work... how about:

    :ref:`west tool <get_the_code>`.

dbkinder commented on doc/release-notes.rst:
```diff
@@ -12,12 +12,12 @@ versions of the Zephyr project are released approximately every three-months.
-   git clone https://github.com/zephyrproject-rtos/zephyr
-   cd zephyr
+   
+   cd zephyrproject/zephyr
```
shouldn't the west commands to get the repo be here (replacing the ``git clone`` command)?

dbkinder replied:
There's also mention of cloning the Zephyr repo over in https://www.zephyrproject.org/developers/#downloads.  I take it a simple clone is not going to work any more?

dbkinder replied:
This text is not in GitHub. It’s separately maintained on the Wordpress site. Once we agree on how to get the source code, I can update that site to reference the instructions.

carlescufi replied:
I have now modified this so that it shows clearly that the instructions are valid until 1.13.

dbkinder commented on doc/release-notes.rst:
```diff
@@ -12,12 +12,12 @@ versions of the Zephyr project are released approximately every three-months.
 you can either download source as a tar.gz file (see the bottom of the
 `GitHub tagged releases`_ page corresponding to each release), or use
-Git clone and checkout commands, such as:
+Git checkout commands, on an existing Zephyr installation obtained using the
```
On the line above this, there's mention of using the tar.gz file to get the source code.  Will that still work?  @nashif

carlescufi replied:
This won't work directly, no. I've modified this now to reflect reality.

carlescufi commented:
@SebastianBoe @nashif @dbkinder  addressed your comments
@nashif we need to add `pip3 install west` to docker in order for this PR to work

@dbkinder @nashif @mbolivar we need to think about how to document releases, should we tag the manifest repo?

mbolivar commented:
> @dbkinder @nashif @mbolivar we need to think about how to document releases, should we tag the manifest repo?

We mostly have a plan for this as discussed in https://github.com/zephyrproject-rtos/west/issues/39#issuecomment-423629912. The short story is:

1. Tag every project (zephyr gets v1.14.0, hypothetical-other-thing gets v1.14.0, etc.)
2. Set the revisions for each project in the manifest file to v1.14.0; tag the commit that does that v1.14.0 as well

Then you can do `west init --mr v1.14.0`. But there is a problem: west itself will continue to be updated, which is not what we want, as the v1.14-compatible version of west should probably stay fixed.

We need to think this through; I've created https://github.com/zephyrproject-rtos/west/issues/106 for that.

mbolivar commented on doc/conf.py:
```diff
@@ -26,11 +27,20 @@
+    west_path = desc.decode(sys.getdefaultencoding()).strip()
+    # Add west, to be able to pull in its API docs.
+    sys.path.append(os.path.join(west_path, 'src'))
+    # HACK: also add the runners module, to work around some import issues
```
No longer needed. Making everything a subpackage of `west` solved this problem.

carlescufi replied:
I could've sworn I had tested it without this hack and it failed. But I clearly didn't since I've now removed it and it works. Thanks.

mbolivar commented on doc/conf.py:
```diff
@@ -26,11 +27,20 @@
+    # related to west's current packaging.
+    sys.path.append(os.path.join(west_path, 'src', 'west'))
+except CalledProcessError as e:
+    # Do nothing, there will be warnings during the build
```
Are you sure about this? We won't be able to build the docs for west's runners; why not fail fast since documentation takes a long time to build especially on slow machines?

carlescufi replied:
Yes good point. I've fixed this now.

mbolivar commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,60 @@ Follow one of the following guides for your host operating system.
-   git clone https://github.com/zephyrproject-rtos/zephyr
+   mkdir zephyrproject
+   cd zephyrproject
+   west init .
```
the directory defaults to cwd, so `west init` will suffice

carlescufi replied:
Yes I know, but I thought this would highlight the fact that one can specify the installation folder. I originally wrote `west init zephyrproject` but changed it to what it is now in order to avoid confusing people, who could perhaps not see at first sight that `zephyrproject` is just a folder name and not a special name you pass to `west` in order to clone Zephyr and not some other project.

mbolivar replied:
I think people will see this and assume the directory is mandatory, and I'm eager to minimize the number of "requirements" (perceived or real) needed to install zephyr.

carlescufi replied:
In that case, and if you want to minimize the number of steps too, should we have:
```
west init zephyrproject
```
instead?

mbolivar commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,60 @@ Follow one of the following guides for your host operating system.
+   west init .
+   west clone
+
+This will clone West itself, along with any projects required by Zephyr. See
```
dangling "see"

carlescufi replied:
Thanks, fixed

mbolivar commented on doc/release-notes.rst:
```diff
@@ -9,17 +9,16 @@ Zephyr project is provided as source code and build scripts for different
+With the introduction of the :ref:`west` tool after the release of Zephyr 1.13,
+it is no longer recommended to download or clone the source code manually.
+Instead we recommend to follow the instructions in :ref:`get_the_code` to do
+so with the help of West.
```
your past self is asking you to not capitalize west ;)

carlescufi replied:
All addressed

mbolivar commented on doc/west/index.rst:
```diff
@@ -1,14 +1,75 @@
+####
 
-The Zephyr project is developing a swiss-army knife command line tool
+The Zephyr project maintains a swiss-army knife command line tool
```
s/maintains/includes/?

carlescufi replied:
Yep, better :)

mbolivar commented on doc/west/index.rst:
```diff
@@ -1,14 +1,75 @@
+full set of functionality offered by the tool:
+
+* Bootstrapper: Installed using ``pip``, implements ``west init``
+* Installation: Installed using ``west init``, implements all other commands
```
I struggle with finding a good name for this but I'm not sure "installation" is right.

How about "main" west?

carlescufi replied:
I went over this several times and I thought this was the least bad one because of the following reasons:

* We are already using "installation" in some other parts of the doc, even before west
* It is unambiguous and includes both the fact that west is installed but also Zephyr itself in its term
* We might choose to add toolchain bits as part of the manifest in the future
* The other options have real drawbacks from my point of view:
  - "Worskpace" could suggest "view" in the model-view-controller metaphor, and this is more than just the view
  - "Main west" doesn't include the projects or manifest, just west itself. I am looking for a term that I can use both for the part of west that is cloned by `west init` but also that can be used generically to refer to the whole contents of the folder created with it

Let's discuss it a bit more.

mbolivar commented on doc/west/index.rst:
```diff
@@ -1,14 +1,75 @@
+first in order to use the rest of functionality included in ``west``, since
+``west init`` will do the following:
+
+* Clone West itself in a :file:`west/west` folder
```
s/West/west/

mbolivar replied:
I'll leave it to you to grep for the rest

carlescufi replied:
done

mbolivar commented on doc/west/index.rst:
```diff
@@ -1,14 +1,75 @@
+Additionally, and once ``west clone`` has been run, the installation will also
+contain a clone of any repositories listed in the manifest. The directory tree
+layout of an installation using the default Zephyr manifest therefore looks
+like this:
```
Worth mentioning the config file here too

carlescufi replied:
done

mbolivar replied:
Here's an updated tree as well:

```
├── west
│   ├── config
│   ├── manifest
│   └── west
└── zephyr
```

carlescufi replied:
I purposefully avoided having `config` in that tree. The reason is that I only wanted to show folders and not files, since I describe it as "directory tree", and I did not want people to think that `config` was a folder. Any suggestions?

mbolivar reviewed (changes requested):
Thanks so much @carlescufi for getting this started. I've got a few nits, but this is looking really good already.

carlescufi commented:
@mbolivar I've addressed your comments
@dbkinder @mbolivar @nashif I've added a section on history and motivation behind west

mbolivar commented on doc/west/index.rst:
```diff
@@ -1,14 +1,122 @@
+To learn more about west's handling of multiple repositories refer to
+:ref:`west-multi-repo`.
+
+All west commands except ``west init`` are implemented by the :file:`west/west`
```
Much clearer, thanks.

I could just be hearing cacofonía where there isn't any, so feel free to ignore, but what about:

```
The west bootstrapper delegates handling for all commands except ``init`` to the installation in :file:`west/west`.
```

carlescufi replied:
Nice, I like it thanks!

carlescufi commented on doc/west/index.rst:
```diff
@@ -1,14 +1,122 @@
+Multiple Git Repositories
+=========================
+
+Zephyr intends to provide all required building blocks needed to deply complex
```
Fixed

nashif committed:
ci: experimenting with west clones

Would be great if west did not modify branch in existing clones.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

mbolivar reviewed (dismissed):
LGTM

nashif commented:
> @nashif can you please remove the cloning instructions from [zephyrproject.org/developers/#downloads](https://www.zephyrproject.org/developers/#downloads) ?

done

tejlmand commented on doc/conf.py:
```diff
@@ -26,11 +27,19 @@
+    # Add west, to be able to pull in its API docs.
+    sys.path.append(os.path.join(west_path, 'src'))
+except FileNotFoundError as e:
+    raise RuntimeError('Unable to find west. Makes sure you have installed'
```
I guess, it is not enough to have installed `west` using pip here.

I guess the requirement is to actually have cloned Zephyr using `west init`, correct ?

carlescufi replied:
Correct, fixed

tejlmand reviewed (changes requested):
I guess, it is not enough to have installed west using pip here.

I guess the requirement is to actually have cloned Zephyr using west init, correct ?

tejlmand reviewed (dismissed):
Looking good.

tejlmand reviewed (approved):
approved

pfalcon commented:
@carlescufi : Was it the intention to merge this into topic-west branch? I thought the idea to merge it into master, to let (make) users use up to date west?

carlescufi commented:
@pfalcon yes, we want to collect all required changes in topic-west first, then send an email to the mailing list and at the same time merge topic-west into master
